src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -221,32 +238,45 @@ private async Task UpdateSourceSymbolTreeInfoAsync(Project project, Cancellation
 
                     // Mark that we're up to date with this project.  Future calls with the same 
                     // semantic version can bail out immediately.
-                    _projectToInfo.AddOrUpdate(project.Id, projectInfo, (_1, _2) => projectInfo);
+                    _projectToInfo[project.Id] = projectInfo;
                 }
             }
 
-            private Task GetTask(Project project, Func<Task> func, CancellationToken cancellationToken)
-            {
-                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
-                return isRemoteWorkspace
-                    ? Task.Run(func, cancellationToken)
-                    : func();
-            }
-
-            private Task UpdateReferencesAync(Project project, CancellationToken cancellationToken)
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/33172", AllowGenericEnumeration = false, AllowCaptures = false)]
+            private Task UpdateReferencesAsync(Project project, CancellationToken cancellationToken)
             {
                 // Process all metadata references. If it remote workspace, do this in parallel.
-                var tasks = new List<Task>();
+                var tasks = new List<Task>(project.MetadataReferences.Count);
+                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
 
-                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
+                // MetadataReferences is stored as ImmutableArray, but exposed as IReadOnlyList
+                // so cast cheaply to avoid boxed enumerator.
+                foreach (var metadataReference in project.MetadataReferences.ToImmutableArrayOrEmpty())
```

Comment:
❗️ Document this with an assertion:

```csharp
Debug.Assert(project.MetadataReferences is ImmutableArray<MetadataReference>, "We rely on specific collection types to ensure allocations are avoided.");
```

Reply:
I'm also hugely surprised this was a big thing that needed to be optimized.  There should only be some finite N number of projects... and we're just doing this operation across each one of them.  

If this really was showing up... it feels to me liek the issue might be elsewhere (i.e. something calling this too many times) versus this method actually being the one at fault.

Reply:
Project count in the solution I tested was around 200.

Reply:
Added assert.

@CyrusNajmabadi I've went and cleaned up most of the allocations I could find on this path when I saw it come up while typing.

Maybe there's something wrong in the VSMac host, WRT checksums, but I've done validation recently and they seem okay.